components/actions_menu/actions_menu.tsx

```diff
@@ -0,0 +1,443 @@
+
+        // const isMobile = this.props.isMobileView TODO;
+
+        const pluginItems = this.props.pluginMenuItems?.
+            filter((item) => {
+                return item.filter ? item.filter(this.props.post.id) : item;
+            }).
+            map((item) => {
+                if (item.subMenu) {
+                    return (
+                        <Menu.ItemSubMenu
+                            key={item.id + '_pluginmenuitem'}
+                            id={item.id}
+                            postId={this.props.post.id}
+                            text={item.text}
+                            subMenu={item.subMenu}
+                            action={item.action}
+                            root={true}
+                        />
+                    );
+                }
+                return (
+                    <Menu.ItemAction
+                        key={item.id + '_pluginmenuitem'}
+                        text={item.text}
+                        onClick={() => {
+                            if (item.action) {
+                                item.action(this.props.post.id);
+                            }
+                        }}
+                    />
+                );
+            }) || [];
+
+        let appBindings = [] as JSX.Element[];
+        if (this.props.appsEnabled && this.state.appBindings) {
+            appBindings = this.state.appBindings.map((item) => {
+                let icon: JSX.Element | undefined;
+                if (item.icon) {
+                    icon = (
+                        <img
+                            key={item.app_id + 'app_icon'}
+                            src={item.icon}
+                        />);
+                }
+
+                return (
+                    <Menu.ItemAction
+                        text={item.label}
+                        key={item.app_id + item.location}
+                        onClick={() => this.onClickAppBinding(item)}
+                        icon={icon}
+                    />
+                );
+            });
+        }
+
+        const {formatMessage} = this.props.intl;
+
+        const divider = (
+            <li
+                className='MenuItem__divider'
+                role='menuitem'
+            />);
+
+        let marketPlace = null;
+        if (this.props.isSysAdmin) {
+            marketPlace = (
+                <>
+                    {divider}
+                    <Menu.ItemAction
+                        id={`marketplace_icon_${this.props.post.id}`}
+                        show={true}
+                        text={formatMessage({id: 'post_info.marketplace', defaultMessage: 'App Marketplace'})}
+                        icon={Utils.getMenuItemIcon('icon-view-grid-plus-outline')}
+                        onClick={this.handleOpenMarketplace}
+                    />
+                </>
+            );
+        }
+
+        if (isSystemMessage) {
```

Comment:
Does it make sense to get this near the beginning of the function?

Reply:
yes. done.